lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,329 @@
+  };
+  template<typename T>
+  bool IsLessThanOrEqualTo(const T &strVal, const T &upperStrVal) {
+    return (strVal.compare(upperStrVal) <= 0);
+  }
+  bool IsLessThanOrEqualTo(
+      const bool &logicalVal, const bool &upperLogicalVal) {
+    return logicalVal == upperLogicalVal;
+  }
+  bool IsLessThanOrEqualTo(
+      const std::int64_t &intVal, const std::int64_t &upperIntVal) {
+    return intVal <= upperIntVal;
+  }
+  template<typename T>
+  bool CheckOverlappingValue(
+      std::optional<T> lowerVal, std::optional<T> upperVal) {
+    return (!lowerVal || !upperVal ||
+        IsLessThanOrEqualTo(lowerVal.value(), upperVal.value()));
+  }
+  // [a, b] & [c, d] overlaps if: (c<=b && a<=d)
+  template<typename T>
+  bool CheckOverlappingRangeValue(std::optional<T> memberFromVal,
+      std::optional<T> memberToVal, std::optional<T> rangeFromVal,
+      std::optional<T> rangeToVal) {
+    return (CheckOverlappingValue(memberFromVal, rangeToVal) &&
+        CheckOverlappingValue(rangeFromVal, memberToVal));
+  }
+  template<typename T>
+  bool IsRangeValueOverlapping(SelectCaseRangeType<T> &rangeValue,
+      parser::CharBlock &overlappingSource) {
+    auto &rangeValsHandle{GetRangeValHandle<T>()};
+    auto selectCaseRangeValsIter{std::find_if(rangeValsHandle.begin(),
+        rangeValsHandle.end(), [&rangeValue, this](const auto &mem) {
+          return ((!mem.fromVal && !rangeValue.fromVal) ||
+              (!mem.toVal && !rangeValue.toVal) ||
+              this->CheckOverlappingRangeValue(mem.fromVal, mem.toVal,
+                  rangeValue.fromVal, rangeValue.toVal));
+        })};
+    if (selectCaseRangeValsIter != rangeValsHandle.end()) {
+      overlappingSource = selectCaseRangeValsIter->source;
+      return true;
+    }
+    return false;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, bool>::value,
+      std::vector<SelectCaseRangeType<bool>> &>::type
+  GetRangeValHandle() {
+    return selectCaseLogicalRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::int64_t>::value,
+      std::vector<SelectCaseRangeType<std::int64_t>> &>::type
+  GetRangeValHandle() {
+    return selectCaseIntRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::string>::value,
+      std::vector<SelectCaseRangeType<std::string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseStringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u16string>::value,
+      std::vector<SelectCaseRangeType<std::u16string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU16StringRangeVals_;
+  }
+  template<typename T>
+  typename std::enable_if<std::is_same<T, std::u32string>::value,
+      std::vector<SelectCaseRangeType<std::u32string>> &>::type
+  GetRangeValHandle() {
+    return selectCaseU32StringRangeVals_;
+  }
+
+  std::vector<SelectCaseRangeType<bool>> selectCaseLogicalRangeVals_;
```

Comment:
This representation allows a `SELECT CASE` construct to have cases of all of these distinct types, but that's more complicated than Fortran requires.  All of the cases in a `SELECT CASE` construct will have the same type, which is the type on which `SelectCaseHelper` should be templatized.

Reply:
Yes, the right approach would be to make the class SelectCaseHelper templatized. Like below:

template<typename T>
class SelectCaseHelper {
public:
  SelectCaseHelper(parser::Messages &messages, int kind, TypeCategory category)
    : kind_{kind}, category_(category), messages_{messages} {}
  ...
private:
  struct SelectCaseRangeType { ... };
  ...
  std::vector<SelectCaseRangeType> selectCaseRangeVals_;
  int kind_;
  TypeCategory category_;
  parser::Messages &messages_;
};

This way only one vector data member is needed and we do not need helper functions
to get the handle as well.

Then while creating the object, we should instantiate with the exact type. Like below:
SelectCaseHelper<bool> selectCaseStmts
or
SelectCaseHelper<std::int64_t> selectCaseStmts
or
SelectCaseHelper<std::string> selectCaseStmts
or so on

All we have is the KIND and CATEGORY of the expression type while creating the object.
This is where I got stuck for a long time. I tried to use factory method by creating a dummy base class
for SelectCaseHelper and tried to instantiate the SelectCaseHelper class with specific types based on the 
KIND and CATEGORY values. This method does not work if we have a member function template. 

With current implementation, vector object of all types co-exists, but only one will be used based on
the SELECT CASE expression type, if there is a mismatch of type or kind, we anyway issue errors for type
and kind mismatch.

This is the working soulution I could get, hence I added a TODO to make this class templatized as
you had mentioned earlier that you will implement a way to extract the exact type given the KIND and CATEGORY.
Once that is in place I can make the class templatized and remove redundant vector data members,
corresponding helper functions.

Kindly suggest if you know a way to extract the exact type to instantiate the class SelectCaseHelper.

Reply:
> This representation allows a SELECT CASE construct to have cases of all of these distinct types, but that's more complicated than Fortran requires.  All of the cases in a SELECT CASE construct will have the same type, which is the type on which SelectCaseHelper should be templatized.
Hi @klausler , 
I have converted SelectCaseHelper to a template class, there by redundant data member and supporting members functions have been removed.